File: Notepad++/网易笔试：喝咖啡.cpp
#include <iostream>
#include <vector>
using namespace std;

int func(vector<int> test, int k, int m)
{
	int count = m;
	if(m == 0) return int(30 / (k + 1));
	vector<int> x, y, z;
	int mid;
	int v = test[0];
	while(v > 0)
	{
		if(v - k - 1 >= 1)
		{
			x.push_back(v - k - 1);
			count += 1;
		}
		v -= k + 1;
	}
	
	int back = test[m - 1];
	while(back < 30)
	{
		if(back + k + 1 <= 30)
		{
			y.push_back(back + k + 1);
			count += 1;
		}
		back += k + 1;
	}
	
	for(int i = 0; i < m - 1; ++i)
	{
		mid = test[i] + k + 1;
		while(mid + k + 1 <= test[i + 1])
		{
			z.push_back(mid + k + 1);
			count += 1;
			mid += k + 1;
		}
	}
	return count;
}

int main()
{
	int n;
	cin >> n;
	vector<int> test;
	for(int i = 0; i < n; ++i)
	{
		int k, m, d;
		cin >> k >> m;
		test.clear();
		for(int i = 0; i < m; ++i)
		{
			cin >> d;
			test.push_back(d);
		}
		cout << func(test, k, m) << endl;
	}
	return 0;
}

File: Notepad++/网易笔试：喝咖啡.py
#

k = 1
m = 15
test = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29]

k = 1
m = 7
test = [5, 9, 13, 17, 21, 25, 29]

k = 3
m = 3
test = [5, 10, 20]

if len(test) == 0:
    count = 30 // (k + 1)
    print(count)
else:
    w = []

    count = len(test)

    v = test[0]
    while v > 0:
        if v - k - 1 >= 1:
            w.append(v - k - 1)
            count += 1
        v -= k + 1

    # print(w + test)

    ww = []
    back = test[-1]
    while back < 30:
        if back + k + 1 <= 30:
            ww.append(back + k + 1)
            count += 1
        back += k + 1

    www = []
    for i in range(0, len(test) - 1):
        mid = test[i] + k + 1
        while mid + k + 1 <= test[i + 1]:
            www.append(mid)
            count += 1
            mid += k + 1

    # print(w + test + ww)
    print(w, ww, www)
    print(count)